### PR TITLE
Added "--public" flag to publish public images

### PR DESCRIPTION
**Pull request by mattcui**
@maximilien Synchronized Godeps with master, and re-submitted PR, please review. Thanks.

### CODE REVIEW

**Comment by maximilien** [comment]
@mattcui, what's the story for this PR? I cannot review. Please add story and I will review.

**Comment by mattcui** [comment]
https://www.pivotaltracker.com/story/show/118897023

Already updated the tracker with this submission.

**Comment by maximilien** [comment]
@mattcui that's not the point. Every PR needs to reference tracker, so in your comment for the PR, please include [#118897023](https://www.pivotaltracker.com/story/show/118897023). As we agreed.

File: cmds/import_image/import_image_cmd.go [hunk 1]
```diff
@@ -4,20 +4,47 @@ import (
...
+	boshretry "github.com/cloudfoundry/bosh-utils/retrystrategy"
 	sldatatypes "github.com/maximilien/softlayer-go/data_types"
-	softlayer "github.com/maximilien/softlayer-go/softlayer"
+	"github.com/maximilien/softlayer-go/softlayer"
+	"github.com/pivotal-golang/clock"
+	"time"
+)
+
+var (
+	locations = []sldatatypes.SoftLayer_Location{
```
Comment (maximilien) [inline cmds/import_image/import_image_cmd.go:17]: Why are the locations fixed in code. Add them to a JSON file that can be loaded or passing them in some flag, e.g., `--public-locations-file`.

If these are defaults then they need to be Golang constants and used only if the --public-locations-file` is not used.

File: cmds/import_image/import_image_cmd.go [hunk 2]
```diff
@@ -84,5 +114,28 @@ func (cmd *ImportImageCmd) Run() error {
...
+					return true, errors.New(fmt.Sprintf("There would be an active transaction in progress."))
+				}
+
+				cmd.Id = id
+				cmd.Uuid = ""
+
+				return false, nil
+			})
+		timeService := clock.NewClock()
+		timeoutRetryStrategy := boshretry.NewTimeoutRetryStrategy(5*time.Minute, 30*time.Second, execStmtRetryable, timeService, boshlog.NewLogger(boshlog.LevelInfo))
```
Comment (maximilien) [inline cmds/import_image/import_image_cmd.go]: put these constants as Golang constants so we know easily where to change them

**Comment by maximilien** [comment]
Please address my comments and I'll review quickly and merge (if all addressed).

File: cmds/import_image/import_image_cmd.go [hunk 3]
```diff
@@ -84,5 +114,28 @@ func (cmd *ImportImageCmd) Run() error {
...
 
+	if cmd.Public {
+		execStmtRetryable := boshretry.NewRetryable(
+			func() (bool, error) {
+				id, err := vgbdtgService.CreatePublicArchiveTransaction(cmd.Id, cmd.PublicName, cmd.PublicNote, cmd.PublicNote, locations)
+				if err != nil {
+					return true, errors.New(fmt.Sprintf("There would be an active transaction in progress."))
+				}
+
+				cmd.Id = id
```
Comment (maximilien) [inline cmds/import_image/import_image_cmd.go:114]: I don't understand these two lines. Why is the `cmd.Uuid` set to `""`

File: cmds/import_image/import_image_cmd_test.go [hunk 4]
```diff
@@ -140,7 +140,13 @@ var _ = Describe("import-image command", func() {
 			}
 		})
 
-		It("creates a VGDTG with UUID and ID", func() {
+		It("creates a private VGDTG with UUID and ID", func() {
```
Comment (maximilien) [inline cmds/import_image/import_image_cmd_test.go:144]: Add a negative test (at least).

File: cmds/import_image/import_image_cmd.go (hunk 1 above)
Comment (mattcui) [inline cmds/import_image/import_image_cmd.go:17]: They should be regarded as initial locations where the public image will be stored, if a user needs to use the image in a different location, the image should be transferred to that location automatically. So I don't think it's necessary to add one more flag.

File: cmds/import_image/import_image_cmd.go (hunk 2 above)
Comment (mattcui) [inline cmds/import_image/import_image_cmd.go]: Seems array can't be declared as a constant, got the following error if I declare it a constant:

```
cmds/import_image/import_image_cmd.go:34: const initializer []data_types.SoftLayer_Location literal is not a constant
```

File: cmds/import_image/import_image_cmd.go (hunk 3 above)
Comment (mattcui) [inline cmds/import_image/import_image_cmd.go:114]: The CreatePublicArchiveTransaction function only returns id of the image, no uuid. Actually, eCPI only uses image id to deploy VMs, so uuid is not required, if we want to also add uuid in the light stemcell for public images, we need Softlayer-go to support a new API [getGlobalIdentifier](http://sldn.softlayer.com/reference/services/SoftLayer_Virtual_Guest_Block_Device_Template_Group/getGlobalIdentifier).

**Comment by mattcui** [comment]
@maximilien since there imported 2 new 3rd-party repositories (see below), need to add them into Godeps, and seems there is new version of ginkgo/gomega to be updated in Godeps, there will be a lots of changes on Godeps in the PR, what's your advice on how to handle it? Thanks.

```
github.com/cloudfoundry/bosh-utils
github.com/pivotal-golang/clock
```

File: cmds/import_image/import_image_cmd_test.go (hunk 4 above)
Comment (mattcui) [inline cmds/import_image/import_image_cmd_test.go:144]: Added, please help review. Thanks.

File: cmds/import_image/import_image_cmd.go (hunk 3 above)
Comment (mattcui) [inline cmds/import_image/import_image_cmd.go:114]: I also confirmed that uuid will be retrieved in create_light_stemcell task, so uuid is not needed to set in import_stemcell here.

**Comment by maximilien** [comment]
@mattcui Let me do Godeps changes. We do not need to always be on the bleeding edge of Ginkgo and Gomega. Please do not update these dependencies until you need to.

File: cmds/import_image/import_image_cmd.go (hunk 3 above)
Comment (maximilien) [inline cmds/import_image/import_image_cmd.go:114]: Add a issue in `softlayer-go` for `getGlobalIdentifier` and I will address it when I have a minute.

File: cmds/import_image/import_image_cmd.go [hunk 5]
```diff
@@ -4,20 +4,46 @@ import (
...
 
+	boshlog "github.com/cloudfoundry/bosh-utils/logger"
+	boshretry "github.com/cloudfoundry/bosh-utils/retrystrategy"
 	sldatatypes "github.com/maximilien/softlayer-go/data_types"
-	softlayer "github.com/maximilien/softlayer-go/softlayer"
+	"github.com/maximilien/softlayer-go/softlayer"
+	"github.com/pivotal-golang/clock"
+)
+
+var (
```
Comment (maximilien) [inline cmds/import_image/import_image_cmd.go:16]: Where is the change to move this to a Golang `const`?

File: cmds/import_image/import_image_cmd_test.go [hunk 6]
```diff
@@ -153,5 +160,72 @@ var _ = Describe("import-image command", func() {
...
+			common.TIMEOUT = 10 * time.Millisecond
+			common.POLLING_INTERVAL = 1 * time.Millisecond
+
+			err = cmd.Run()
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(importImageCmd.Uuid).To(Equal(""))
+			Expect(importImageCmd.Id).ToNot(Equal(""))
+		})
+		It("creates a public VGDTG with UUID and ID - Nagtive path", func() {
```
Comment (maximilien) [inline cmds/import_image/import_image_cmd_test.go]: remove `"Nagtive path"` (type in first word) and move this to a `Context("...")`

`Context` are meant exactly to divide tests into logical sections. So something like:

```
Context("when <state of the world is bad>", funct() {
  It("fails to do xyz", func() {
  })

  It("fails to do abc", func() {
  })
})

Context("when <state of the world is happy>", funct() {
  It("succeeds doing xyz", func() {
  })

  It("succeeds doing abc", func() {
  })
})

```

File: cmds/import_image/import_image_cmd.go (hunk 3 above)
Comment (mattcui) [inline cmds/import_image/import_image_cmd.go:114]: I raised [softlayer-go issue 103](https://github.com/maximilien/softlayer-go/issues/103) there. But according to my last comment, create_light_stemcell task retrieves all properties for the public image, including uuid. So actually it's not needed to get uuid in import_stemcell task.

File: cmds/import_image/import_image_cmd_test.go (hunk 6 above)
Comment (mattcui) [inline cmds/import_image/import_image_cmd_test.go]: Made the changes on UT, please help review, thanks.

File: cmds/import_image/import_image_cmd.go (hunk 5 above)
Comment (mattcui) [inline cmds/import_image/import_image_cmd.go:16]: Let's just set lon02 as an initial DC, same as private image we created currently, we don't see any problem so far, we could enhance it later as you suggested to specify a DC list in a json file or command option. OK?

**Comment by mattcui** [comment]
@maximilien  I think I already addressed all of your comments, please help review code again. Thanks.

File: cmds/import_image/import_image_cmd.go (hunk 5 above)
Comment (maximilien) [inline cmds/import_image/import_image_cmd.go:16]: why not do it now? Let's talk at scrum. If you convince me then I'll merge

Comment (mattcui) [inline cmds/import_image/import_image_cmd.go:16]: Do what? you want to publish the image to all DCs? We already tested that we only published the image into lon02 DC, and we tried to deploy a VM with this image in dal09 DC, it's successful, so I don't think it's necessary to populate the image to all DCs.

Comment (mattcui) [inline cmds/import_image/import_image_cmd.go:16]: I already updated the code to only set lon02 there.

**Comment by maximilien** [comment]
I still want to see an option for passing a JSON file for the various locations. Also, I would like to see more tests. I am accepting but also adding another story [#120140843]  for these missing parts.

**Pull request merged**